File: 001-500/0041. First Missing Positive.java
// Hard
// In-place Hashing
// Time:O(n),Space:O(1)
// https://leetcode.cn/problems/first-missing-positive/

// To reach O(1) space, we need to do in original arr.
// For this question, we actually need to check if it appears before, so need a way to mark if it appears.
// We can use in-place Hashing -> serve index as key, and put it's value in the "correct" place.

class Solution {
    public int firstMissingPositive(int[] nums) {
        int n = nums.length;

        for (int i = 0; i < n; i++) {
            // nums[i] should put in nums[nums[i] - 1]
            while (nums[i] > 0 && nums[i] <= n && nums[nums[i] - 1] != nums[i]) {
                int tmp = nums[nums[i] - 1];
                nums[nums[i] - 1] = nums[i];
                nums[i] = tmp;
            }
        }

        // Find the first place with wrong number
        for (int i = 0; i < n; i++) {
            if (nums[i] != i + 1)
                return i + 1;
        }
        return n + 1; // The case we actually perfectly match with all numbers, like [1,2,3...n]
    }
}

File: 0001-0500/0041. First Missing Positive.py
# Hard
# In-place Hashing
# Time:O(n), Space:O(1)
# https://leetcode.cn/problems/first-missing-positive/

from typing import *

class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        n = len(nums)

        # put number in the "correct" position
        # number i should be placed in index i-1
        for i in range(n):
            while 1 <= nums[i] <= n and nums[nums[i] - 1] != nums[i]:
                # swap
                exp_idx = nums[i] - 1
                nums[i], nums[exp_idx] = nums[exp_idx], nums[i]
        
        # find the number not place in the correct place
        for i in range(n):
            if nums[i] != i + 1: # nums[i] should place in i+1
                return i + 1
        
        # if all in the correct place
        return n + 1
    